mrmustard/physics/ansatze.py

```diff
@@ -0,0 +1,351 @@
+            return
+        indices_to_check = set(range(self.batch_size))
+        removed = []
+        while indices_to_check:
+            i = indices_to_check.pop()
+            for j in indices_to_check.copy():
+                if np.allclose(self.mat[i], self.mat[j]) and np.allclose(self.vec[i], self.vec[j]):
+                    self.array = math.update_add_tensor(self.array, [i], [self.array[j]])
+                    indices_to_check.remove(j)
+                    removed.append(j)
+        to_keep = [i for i in range(self.batch_size) if i not in removed]
+        self.mat = math.gather(self.mat, to_keep, axis=0)
+        self.vec = math.gather(self.vec, to_keep, axis=0)
+        self.array = math.gather(self.array, to_keep, axis=0)
+        self._simplified = True
+
+    def simplify_v2(self) -> None:
+        r"""A different implementation that orders the batch dimension first."""
+        if self._simplified:
+            return
+        self._order_batch()
+        to_keep = [d0 := 0]
+        mat, vec = self.mat[d0], self.vec[d0]
+        for d in range(1, self.batch_size):
+            if np.allclose(mat, self.mat[d]) and np.allclose(vec, self.vec[d]):
+                self.array = math.update_add_tensor(self.array, [d0], [self.array[d]])
+            else:
+                to_keep.append(d)
+                d0 = d
+                mat, vec = self.mat[d0], self.vec[d0]
+        self.mat = math.gather(self.mat, to_keep, axis=0)
+        self.vec = math.gather(self.vec, to_keep, axis=0)
+        self.array = math.gather(self.array, to_keep, axis=0)
+        self._simplified = True
+
+    def _order_batch(self):
+        generators = [
+            itertools.chain(
+                math.asnumpy(self.vec[i]).flat,
+                math.asnumpy(self.mat[i]).flat,
+                math.asnumpy(self.array[i]).flat,
+            )
+            for i in range(self.batch_size)
+        ]
+        sorted_indices = argsort_gen(generators)
+        self.mat = math.gather(self.mat, sorted_indices, axis=0)
+        self.vec = math.gather(self.vec, sorted_indices, axis=0)
+        self.array = math.gather(self.array, sorted_indices, axis=0)
+
+
+class PolyExpAnsatz(PolyExpBase):
+    """
+    Represents the ansatz function:
+
+        F(z) = sum_i poly_i(z) exp(z^T A_i z / 2 + z^T b_i),
+
+    where each poly_i is a polynomial in z that can be expressed as:
+
+        poly_i(z) = sum_k c^(i)_k z^k,
+
+    with k being a multi-index. The batch of arrays c^(i) are not just array values but can be polynomials
+    of varying order, defined by the terms arr_k z^k for each i. The matrices A_i and vectors b_i
+    are parameters of the exponential terms in the ansatz, and z is a vector of variables.
+
+    Attributes:
+        A (Batch[Matrix]): The list of square matrices A_i
+        b (Batch[Vector]): The list of vectors b_i
+        c (Batch[Tensor]): The array of coefficients for the polynomial terms in the ansatz.
+
+    Example:
```

Comment:
Does this render, or do you need to replace "Example" with ".. code-block"?

Reply:
I have no idea what renders, I never build the docs 🙈